File: LeetCodeContests/138/previous_permutation.cpp
class Solution:
    def prevPermOpt1(self, A):
        lens = len(A)
        mins = lens - 1
        for i in range(lens - 1, -1, -1):
            if A[i] < A[mins]: mins = i
            if A[i] > A[mins]:
                j = i
                for k in range(lens - 1, -1, -1):
                    if A[k] < A[j]:
                        A[k], A[j] = A[j], A[k]
                        return A
        return A

print(Solution().prevPermOpt1([3, 1, 1, 3]))

File: LeetCodeContests/138/previous_permutation.py
class Solution:
    def prevPermOpt1(self, A):
        lens = len(A)
        mins = lens - 1
        while mins > 0 and A[mins] >= A[mins - 1]:
            mins -= 1
        if mins == 0: return A
        j = mins - 1
        while j >= mins and A[j] >= A[mins - 1]:
            j -= 1
        A[mins - 1], A[j] = A[j], A[mins - 1]
        return A


print(Solution().prevPermOpt1([3, 1, 1, 3]))